### PR TITLE
Feat: Lockfile Parser

### PR DESCRIPTION
**Pull request by justshiv**
- [x] Ready for review
- [x] Follows [CONTRIBUTING](CONTRIBUTING.md) rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?
Adds a `parseLockfile` function to parse the lockfile and returns a `PaketLock` of groups indexed by group name. Paket lockfiles are split into separable groups with different settings, restrictions & dependencies.

Example input to `parseLockfile`:
```
IMPORT-TARGETS: FALSE
FRAMEWORK: NET452
NUGET
  remote: https://api.nuget.org/v3/index.json
    FSharp.Charting (0.90.14)
```

Current example output:
```
{
  "main": {
    "name": "main",
    "repositories": {
      "NUGET": [
        "https://api.nuget.org/v3/index.json"
      ]
    },
    "dependencies": [
      {
        "name": "FSharp.Charting",
        "options": {},
        "version": "0.90.14",
        "group": "main",
        "remote": "https://api.nuget.org/v3/index.json",
        "repository": "NUGET",
        "dependencies": []
      }
    ],
    "options": {
      "import-targets": "false",
      "framework": "net452"
    }
  }
}

```

It also adds the skeletons of `parse` which will generate the full tree to return to our plugin. As this is yet to be built out and requires the inclusion of `paket.dependencies` I haven't added tests for this section.

### CODE REVIEW

File: test/index.test.ts [hunk 1]
```diff
@@ -1,7 +1,14 @@
...
+import {readFileSync} from 'fs';
 
 // dummy test to add jest setup
 describe('test', () => {
   it('jest setup', () => {
+    const file = readFileSync('./test/parse-lock/fixtures/specs-nuget-and-github/paket.lock', 'utf8');
+    const parsed = parse(file);
+    console.log(file);
+    console.log(parsed);
+    console.log(JSON.stringify(parsed, null, 2));
```
Comment (justshiv) [inline test/index.test.ts]: This is left over from currently working through testing out the nested dependencies.

File: test/index.test.ts [hunk 2]
```diff
@@ -1,4 +1,3 @@
-
 // dummy test to add jest setup
 describe('test', () => {
```
Comment (lili2311) [inline test/index.test.ts]: can we delete this? Is it needed?

Comment (justshiv) [reply test/index.test.ts]: it is as this is where the index test is going for the first `buildDepTree()`. I'll tell you when `index.test.ts` and `index.ts` are ready for review. Those are last WIP part.

File: lib/lock-parser.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+      } else if (startsWith(line.data, SPECS)) {
+        continue;
+      }
+    } else {
+      const dep = parseDependencyLine(line);
+
+      if (!dep) {
+        continue;
+      }
+      Object.assign(dep.options, context.options);
```
Comment (justshiv) [inline lib/lock-parser.ts]: After zoom with @lili2311 have agreed to stop doing this and to add the file context to the root result object as well and not just to the dependency, that way allowing us to grab any meta for the parsed file directly and not needing to loop through the list of deps. 

cc. @Kirill89

File: lib/lock-parser.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+          // TODO: Should we do something with it?
+          continue;
+        }
+
+        context.options = context.options || {};
+        context.options[optionName.trim().toLowerCase()] = optionValue.trim();
+      }
+    } else if (line.indentation === 1) {
+      if (startsWith(line.data, REMOTE)) {
+        const parts = line.data.split(':');
```
Comment (justshiv) [inline lib/lock-parser.ts]: Splitting by `:` here is dangerous as `REMOTE` is a URL. This causes a bug.

File: lib/lock-parser.ts [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+          // TODO: Should we do something with it?
+          continue;
+        }
+
+        group.options = group.options || {};
+        group.options[optionName.trim().toLowerCase()] = optionValue.trim().toLowerCase();
+      }
+    } else if (line.indentation === 1) {
+      if (startsWith(line.data, REMOTE)) {
+        const remote = line.data.match(/(?:http(s)?:\/\/)?[\w.-]+(?:\.[\w\.-]+)+[\w\-\._~:/?#[\]@!\$&'\(\)\*\+,;=.]+$/);
```
Comment (justshiv) [inline lib/lock-parser.ts]: I spotted a bug in the case that the remote is not a url e.g.
```
GITHUB
  remote: haf/YoLo
    YoLo.fs (2ae49ac885e2f42812edbaa2dbc8d0e872653cab)
```

Our previous implementation just split by `:` which caused this logic to fail for all urls. Now that we're specifically matching for urls we ignore the above case.

Please address & regenerate the affected fixtures. 🙏

**Comment by justshiv** [comment]
It was requested that we add any meta to the main `PaketLock` returned object as discussed here: https://github.com/snyk/snyk-paket-parser/pull/17#discussion_r255164278. 

I have done this by indexing the main return object by groups and then added the meta to each group  as the meta is group specific. If you have any suggestions on how you can further group meta information to send on the root object and you can add it within the scope of this PR please do.

**Comment by justshiv** [comment]
I've found 2 fixtures that we don't handle well:
- `git-and-github` throws an error because there is an indentation of 3 spaces. How do we want to handle this error gracefully?
- `multiple-remotes` has 2 remotes listed before a list of dependencies. We currently ignore the first remote link entirely in this case. We need to address how to handle this.

**Comment by Kirill89** [comment]
Small optional notes:

- I prefer to not call `default` group `main` because what if user will call real group `main`? We will overwrite it?
- We can create array of groups on top level instead of object `name` => `object`. In this case `name` will be optional and we will solve (1).
- Let’s not convert option names to lower case (to save ability to reprint file from parsed data).
- I’m not sure if we need `repositories` on group level (it can’t help us because we have `remote` and `repository` in each dependency - we can easily use it).
- Not sure if we need `group` on dependency level.

File: lib/lock-parser.ts [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+export interface PaketLock {
+  [name: string]: Group;
+}
+
+function parseOptions(optionsString: string): Option {
+  const options = {} as Option;
+
+  for (const option of optionsString.split(', ')) {
```
Comment (orsagie) [inline lib/lock-parser.ts]: change the string we are splitting on to \,[ ]?\

File: lib/lock-parser.ts [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    const optionParts = option.split(': ');
```
Comment (orsagie) [inline lib/lock-parser.ts]: same here: `\:[ ]?\`

File: lib/lock-parser.ts [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+        if (!optionValue) {
+          // TODO: Should we do something with it?
+          continue;
+        }
+
+        group.options = group.options || {};
+        group.options[optionName.trim().toLowerCase()] = optionValue.trim().toLowerCase();
+      }
+    } else if (line.indentation === 1) {
+      if (startsWith(line.data, REMOTE)) {
```
Comment (orsagie) [inline lib/lock-parser.ts]: upercase or lowercase `line.data` too

File: lib/lock-parser.ts (hunk 5 above)
Comment (orsagie) [reply lib/lock-parser.ts]: any reason why we don't do this: `const remote = line.data.substring(REMOTE.length + ':'.length).trim();`

File: test/lock-parser/fixtures/simple-github/out.json [hunk 9]
```diff
@@ -1,188 +1,179 @@
...
+              "options": {},
+              "version": ">= 3.7"
+            }
+          ]
+        },
+        {
+          "name": "FsCheck",
+          "options": {},
+          "version": "2.9",
+                    "remote": "https://www.nuget.org/api/v2",
```
Comment (Kirill89) [inline test/lock-parser/fixtures/simple-github/out.json]: Weird indentation

Comment (Kirill89) [reply test/lock-parser/fixtures/simple-github/out.json]: * not only here

**Pull request merged**